### PR TITLE
feature/entity_chooser

### PR DESCRIPTION
**Pull request by teamagam1**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/teamagam/GimelGimel/pull/72" target="_blank">ReviewNinja

File: app/src/main/assets/Cesium/Cesium.js [hunk 1]
```diff
@@ -153002,8 +153002,8 @@ define('Scene/Scene',[
     }
 
     // pick rectangle width and height, assumed odd
-    var rectangleWidth = 3.0;
-    var rectangleHeight = 3.0;
+    var rectangleWidth = 55.0;
```
Comment (BarAgam) [inline app/src/main/assets/Cesium/Cesium.js:153005]: Hamsa-Hamsa

File: app/src/main/assets/GimelGimel/CesiumEventHandler.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+};
+
+/**
+ * Returns the list of Entities at the provided window coordinates.
+ * The Entities are sorted front to back by their visual order.
+ *
+ * @param {Cartesian2} windowPosition The window coordinates.
+ * @returns {Entity[]} The picked Entities or undefined.
+ */
+GG.EventHandler.prototype.pickEntities = function (windowPosition) {
```
Comment (BarAgam) [inline app/src/main/assets/GimelGimel/CesiumEventHandler.js]: Unused method. should be deleted IMO

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageMapEntitiesViewModel.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        mSelectedModel = model;
+    }
+
+    @Override
+    public void onEntityClick(Entity entity) {
+        mSelectedModel.select(mEntityToMessageHashMap.get(entity));
+    }
+
+    public Entity addMessage(Message message) {
+        PointGeometry point;
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageMapEntitiesViewModel.java]: extract method that returns PointGeometry from Message !fix

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageMapEntitiesViewModel.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        Entity userEntity = createMessagePinEntity(point);
+        mEntityToMessageHashMap.put(userEntity, message);
+        userEntity.setOnClickListener(this);
+        return userEntity;
+    }
+
+    private Entity createMessagePinEntity(PointGeometry pointGeometry) {
+        //Todo: use symbol interface
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageMapEntitiesViewModel.java]: What did you mean?

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessageDetailViewModel.java [hunk 5]
```diff
@@ -35,6 +36,10 @@ private boolean isSelectedMessageOfType(@Message.MessageType String messageType)
         return mSelectedMessageModel.getSelected().getType().equals(messageType);
     }
 
+    public void drawMessageOnMap(MessagesDetailBaseGeoFragment.GeoMessageInterface drawMessageOnMapInterface) {
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessageDetailViewModel.java]: View-model should be unaware of the view.
This interface functionality should be defining at the View-Model tier !fix

Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessageDetailViewModel.java]: organize method arrangement. public method should appear before protected/private !fix

File: app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java [hunk 6]
```diff
@@ -72,6 +73,7 @@
 
     private VectorLayer mSentLocationsLayer;
     private VectorLayer mUsersLocationsLayer;
+    private VectorLayer mReceivedLocationsLayer;
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:76]: We  should move all the layers to be part of the View-model/model and "plug" them to the map on map initialization. 
If we'll do it, we can update the map by manipulating the entities at the view-model/model layers without messing around with with the fragments/activities.
I'm not asking you to change it all now, just for your thoughts. 
Lets talk about it tomorrow and decide on a future refactoring.

File: app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/messags_panel_fragments/MessagesDetailGeoFragment.java [hunk 7]
```diff
@@ -47,7 +47,7 @@ public void gotoLocationClicked() {
 
     @OnClick(R.id.fragment_messages_detail_show_pin_button)
     public void showPinOnMapClicked() {
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/messags_panel_fragments/MessagesDetailGeoFragment.java:49]: Cant this binding be done at the common abstract DetailFragment now ?

File: app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/messags_panel_fragments/MessagesDetailImageFragment.java [hunk 8]
```diff
@@ -72,7 +72,7 @@ public void gotoImageClicked() {
 
     @OnClick(R.id.fragment_image_show_pin_button)
     public void showPinOnMapClicked() {
-        showPinOnMapClicked(mViewModel.getPointGeometry());
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/messags_panel_fragments/MessagesDetailImageFragment.java:75]: Cant this binding be done at the common abstract DetailFragment now ?

File: app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/cesium/CesiumMapView.java [hunk 9]
```diff
@@ -362,4 +370,23 @@ public void run() {
 
         });
     }
-}
\ No newline at end of file
+
+    @Override
+    public void onCesiumEntityClick(String layerId, String entityId) {
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/cesium/CesiumMapView.java:375]: I'd split this method into validation (which throws an exception), entity parsing, and clicking

File: app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/data/entities/AbsEntity.java [hunk 10]
```diff
@@ -63,6 +65,24 @@ public void removeOnEntityChangedListener() {
...
+    @Override
+    public void setOnClickListener(OnClickListener onClickListener) {
+        mOnClickListener = onClickListener;
+    }
+
+    @Override
+    public void removeOnClickListener() {
+        if (mOnClickListener == null) {
+            //No listener attached
+            sLogger.d("removeOnClickListener called with no listener attached");
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/data/entities/AbsEntity.java]: I think that's a warning. unwanted state that needs investigation.

File: app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/data/entities/Entity.java [hunk 11]
```diff
@@ -23,11 +23,27 @@
...
+    void setOnClickListener(OnClickListener clickListener);
+
+    void removeOnClickListener();
+
+    void clicked();
+
     /**
      * An interface needed to be implemented to register as a listener
-     * for entity changes events
+     * for entity changes events. used for map updates.
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/data/entities/Entity.java]: It doesn't have to be map events.

Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/data/entities/Entity.java]: map updates *

File: app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/data/entities/Entity.java [hunk 12]
```diff
@@ -23,11 +23,27 @@
...
      */
-    interface EntityChangedListener {
+    interface EntityChangedListener{
         void onEntityChanged(Entity changedEntity);
     }
+
+    /**
+     * An interface needed to be implemented to register as a listener
+     * for entity clicked events. used for detecting map clicks.
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/data/entities/Entity.java]: It doesn't have to be related to map.

File: app/src/main/assets/GimelGimel/CesiumEventHandler.js (hunk 2 above)
Comment (teamagam1) [inline app/src/main/assets/GimelGimel/CesiumEventHandler.js]: !completed

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageMapEntitiesViewModel.java (hunk 3 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageMapEntitiesViewModel.java]: !fixed

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageMapEntitiesViewModel.java (hunk 4 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageMapEntitiesViewModel.java]: just reminder that in the future we will have more point types so it should call some symbol creator for each type.

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessageDetailViewModel.java (hunk 5 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessageDetailViewModel.java]: !fixed

Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessageDetailViewModel.java]: about the VMMV pattern: we do not have entities model yet. 
I guessed that you will comment below about this topic :).
so just like you have said, I would not change it now but I totally agree with you about our future plans.

File: app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/messags_panel_fragments/MessagesDetailGeoFragment.java (hunk 7 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/messags_panel_fragments/MessagesDetailGeoFragment.java:49]: yes, but i'll than both views must have the same id.
I prefer not to do that.

File: app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/messags_panel_fragments/MessagesDetailImageFragment.java (hunk 8 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/messags_panel_fragments/MessagesDetailImageFragment.java:75]: same

File: app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/cesium/CesiumMapView.java (hunk 9 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/cesium/CesiumMapView.java:375]: !completed

File: app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/data/entities/AbsEntity.java (hunk 10 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/data/entities/AbsEntity.java]: !completed

**Pull request merged**